### DISCUSSION TITLE
Codegen golang struct?

### DISCUSSION THREAD

**Discussion by shettich** [cue-lang/cue#1855] [Q&A] [answered]
I thought I would try to use cue instead of YAML for some simple config/data inputs to a Go program.  I sort of expected:

- define the schema in cue
- populate data/config files
- codegen go package w/ auto-generated structs with the cli
- load the data in go and use the structs

Is that totally wrong? I don't see a way to codegen the go struct defs.  Do people just define matching structs by hand?

**Comment by verdverm** [comment, marked as answer]
You're not totally off here, and there are several ways to achieve what you want.

1: Defining the schema is a great way to start. It helps enforce the incoming configuration so you don't have to validate with hand written code.

2/4: Populate data / config, I assume this means the actual input to your program. You can set yourself up to read CUE, JSON, and/or Yaml. There are a couple of ways of doing this, but in the end you will be able to decode the input into actual Go structs. If you only want to decode yaml / json, you can avoid the CUE loader. If you want to only support CUE for config input, then you can follow this example: https://cuetorials.com/go-api/loading/simple/. If you want to support all three seamlessly, you'll need to do something like this: https://github.com/hofstadter-io/hof/blob/_dev/lib/cuetils/runtime.go#L132 (until the CUE loader supports this out of the box)

3: generating types from CUE definitions... this is the one with several options and complexities

- The CUE team plans to support generating Go structs eventually
- Generally, there are a lot of complexities that are not captured with simple values or definitions, especially if you consider generating types in several languages from the same definition (some discussion here: https://docs.hofstadter.io/code-generation/type-and-structs/)

Ways you can do this:

- write by hand, and wait for upstream support
- `cue cmd` + `text/template`, you will have to craft templates, but note, the input needs be concrete, fields like `foo: int` will not work
- `hof gen` is the previous, purpose built for generating types and more from CUE, here is an example with Go types: https://docs.hofstadter.io/first-example/data-layer/generating-types/#type-template


There is a #codegen channel in Slack and `hof` has its own Slack as well

**Comment by shettich** [comment]
Ok, thanks.  Thinking more, I don't actually need a Go struct in this case, if I could somehow just walk the tree and work with each field of the (nested) objects.

It seems like the way to do that is:

- use load.Instances()
- then BuildInstance()
- then Validate()
(at this point formatting the object as a string looks fine)
- then iterate via Values()
here I can't figure it out.  Should I be able to use the iterator to check each field, and then Values() again if I find a struct/list? I seem to always nil for it.Value()

**Comment by chivalryq** [comment]
We did something similar before. Here is the case:
1. we have a lot cue struct like:
```cue
parameter: {
  replica: int
}
```
2. We want:
```go
type Parameter struct{
  Replica int `json:"replica"`
}
```
3. We build a sub-command of our CLI for generating struct above: `vela def gen-api` Here is the PR: https://github.com/kubevela/kubevela/pull/3699
4. It's incomplete. We have one problem: Go struct can not handle union type like:
```cue
parameter: {
  args: string | #SecretFrom
}
#SecretFrom:{
  name: string
  namespace: string
  key: string
}
```

5. Generated type don't have any validation method. Which I think should be generated.

**Comment by eadlam** [comment]
I do this by leaning on the protobuf integration:

- define the datatypes in protobuf
- generate cue types from proto types
- Write configs in cue
- Export cue configs to json
- Load json into protobuf

this works quite well.
